server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -286,6 +294,23 @@ func (e *EventChannel) FinalizeInvocation(iid string) error {
 	return nil
 }
 
+// writeCacheStatsToDBWhenReady waits long enough for all apps to flush their
+// cache stats from their in-memory buffers, then updates the invocation cache
+// stats in the DB.
+func (e *EventChannel) writeCacheStatsToDBWhenReady(ctx context.Context, invocationID string) {
+	ctx, cancel := context.WithTimeout(ctx, writeCacheStatsTimeout)
+	defer cancel()
```

Comment:
I think it would be clearer if you set the timeout on the context *before* passing it to writeCacheStatsToDBWhenReady -- think of it as sort of a lease with a timeout and then you give that lease to the goroutine, rather than having the goroutine create one itself

Reply:
Hoisted this piece out of the function, but still creating the context inside the goroutine func so that we can still use the `defer cancel()` idiom. Not ideal, but I think that should at least make it more obvious that the goroutine has a definite timeout?

Reply:
You're right, since these outlive the lifetime of the request they are difficult to cancel. My suggestion missed that!